website/addons/base/views.py

```diff
@@ -201,19 +207,117 @@ def create_waterbutler_log(payload, **kwargs):
     return {'status': 'success'}
 
 
+def get_or_start_render(file_guid, start_render=True):
+    try:
+        file_guid.enrich()
+    except exceptions.AddonEnrichmentError as error:
+        return error.renderable_error
+
+    try:
+        return codecs.open(file_guid.mfr_cache_path, 'r', 'utf-8').read()
+    except IOError:
+        if start_render:
+            # Start rendering job if requested
+            build_rendered_html(file_guid.mfr_download_url, file_guid.mfr_cache_path, file_guid.mfr_temp_path)
+    return None
+
+
 @must_be_valid_project
-def get_waterbutler_render_url(**kwargs):
-    provider = request.args.get('provider')
+def addon_view_or_download_file_legacy(**kwargs):
+    query_params = request.args_to_dict()
+    node = kwargs.get('node') or kwargs['project']
+
+    if kwargs.get('path'):
+        path = kwargs['path']
+    elif kwargs.get('fid'):
+        path = kwargs['fid']
+
+    if kwargs.get('provider'):
+        provider = kwargs['provider']
+    elif 'osffiles' in request.path:
+        provider = 'osfstorage'
```

Comment:
There is no `else` block here, so `provider` may be undefined in the call toe `node.web_url_for`